State improvements: avoidance of duplicate conns, avoid data creep and garbage cleanup of old conns

markterrill opened the pull request:
The key issue is that firebase connections in an app, ie ionic, will often be suspended and then reaped by firestore and/or realtime and this doesn't trigger the firestore-plugin disconnect process to trim the state's connection array. We had hundreds of conns with thousands of emitted events in our test app, I don't want to think what was happening in user's apps.

What does this PR do?

- Effectively upserts new connections rather than blindly adding them
- Changes the emitted event array to a singular value, otherwise it just keeps growing
- Performs garbage cleanup of duplicates if somehow the connections array has gone over 30 items, this is a safety catch and will clean up apps with the prior firestore-plugin version

joaqcid commented:
hi @markterrill thanks for the PR, i have a couple of questions, 
- why did you add the "[cancel previous]" string in `trackBy`?
- the change in the state affects only how we log the connections active?

markterrill commented:
Hi @joaqcid 

- Added the trackby as the easiest informational flag to see in the connection list if any garbage has crept in. An options object would have been nicer but it'd require a fair bit more furniture moving that may not be appreciated.
- Correct, the change of state just changes the logging of active connections. The other issue with the firestore plugin state was the connection emittedAt as an array could grow infinitely long so I changed that to a single value rather than an array of values.

The long list of 'connections' was making it hard to debug an app bug where new realtime events were not being fetched after an app resume. My initial guess was that unnecessary connections were being setup by firebase plugin and it wasn't gluing emitted events together, and when I found the state had hundreds of connections it seemed that guess was right. After adding the new logic so it would upsert a new connection, and the garbage collection for anyone with state persisted, it turned out to be a firebase realtime bug - namely realtime is not resuming connections after an app resume, solution: trigger goOffline/goOnline manually when detecting a background/resume event.

joaqcid commented:
oh ok, i see what you did here, basically changing the log in order to handle firebase issue when reconnecting , is that correct?

regarding `emittedAt` the idea was to track all emissions, but i guess we can just keep last emission... or last 10...

markterrill commented:
Hi @joaqcid 
- I changed the log naming structure, ie appending the cancel previous information, so it was clearer from the logs what were duplicates. It's just extra information that helps anyone looking to know if there are items in there that shouldn't be.
- The emittedAt was previously an infinite length array (ie slowly creeping amount of data, just like a memory leak). Changing it to a single value eliminated that problem and provided the basic information to anyone looking at the state. My logic is that if you are really concerned about the full history you can easily implement your own 'log' within your app by listening to emittedAt and constructing your own array etc.

joaqcid commented:
@markterrill im closing this PR since I included the flag `developmentMode` in the `NgxsFirestoreModule` options to disable logging for production environments, hope this helps!
